src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -117,7 +123,7 @@ internal enum KeepAliveState
         private long _keepAlivePingTimeoutTimestamp;
         private volatile KeepAliveState _keepAliveState;
 
-        public Http2Connection(HttpConnectionPool pool, Stream stream)
+        public Http2Connection(HttpConnectionPool pool, Stream stream, System.Net.Sockets.Socket? socket)
```

Comment:
We don't always have a socket. What do we do if we don't have a socket?

Reply:
If we don't have a `NetworkStream` and a `Socket`, I would assume it's an in-memory connection with an RTT close to zero, so we should fall back to a small static window.

If this assumption does not work for some reason, we'll have to introduce an abstraction for providing RTT estimations.

Reply:
I don't think that's a reasonable assumption in general. 

It seems like another option would be to try to measure RTT ourselves, e.g. by detecting how long it takes to ACK a SETTINGS frame or a PING.

Reply:
That would require a bunch of extra code and changes. To get an accurate estimation, we should send out the PINGs multiple times, on a regular basis. We'll basically end up duplicating the corresponding logic from the TCP stack. (And note that TCP can rely on ACK, and doesn't have to do extra pings)

Unless it blocks some important scenarios, I don't think it's worth to do it now, I would rather handle it as an improvement opportunity for future.

Reply:
If you think that's easier and/or provides better data, then I'm fine with that.

That said, 

(a) We need to work reasonably well for cases where we don't have access to a socket. A very common case here is just a wrapper stream on top of NetworkStream/SslStream that's used for logging or diagnostics, or just for initial connection creation, etc. Ideally this should work just as well as if you don't use ConnectCallback etc. 

It's a really crappy customer story to say "hey we have this cool ConnectCallback, but it will cause a big perf hit in certain cases". We already fixed one major issue here in 6.0 with idle connection polling. Whenever possible, we should strive to *not* assume we are running on top of a Socket.

(b) Getting the RTT times is plat-specific and thus we're going to end up having to do this for Windows, Linux, and MacOS separately. That looks like a fair amount of work, and I'm not sure it's any easier to do all that vs just do our own RTT estimate.

(c) It's not clear to me how accurate we need the RTT estimate to be. Mostly we just want to grow the buffer sufficiently so it doesn't become a bottleneck. If we err on the side of buffering too much, that's probably not a huge problem, as compared to buffering too little.